toolchain/sem_ir/node.h

```diff
@@ -588,269 +171,13 @@ class Node : public Printable<Node> {
-#define CARBON_SEM_IR_NODE_KIND(Name) \
-  using Name = TypedNode<NodeKind::Name, NodeData::Name>;
-#include "toolchain/sem_ir/node_kind.def"
-
-// Implementation details for typed nodes.
-namespace NodeInternals {
-template <typename T>
-using GetHasParseNode = typename T::HasParseNode;
-template <typename T>
-using GetHasType = typename T::HasType;
-
-// Apply Getter<T>, or provide Default if it doesn't exist.
-template <typename T, template <typename> typename Getter, typename Default,
-          typename Void = void>
-struct GetWithDefaultImpl {
-  using Result = Default;
-};
-template <typename T, template <typename> typename Getter, typename Default>
-struct GetWithDefaultImpl<T, Getter, Default, std::void_t<Getter<T>>> {
-  using Result = Getter<T>;
-};
-template <typename T, template <typename> typename Getter, typename Default>
-using GetWithDefault = typename GetWithDefaultImpl<T, Getter, Default>::Result;
-
-// Base class for nodes that have a `parse_node` field.
-struct HasParseNodeBase {
-  Parse::Node parse_node;
-
-  static auto FromParseNode(Parse::Node parse_node) -> HasParseNodeBase {
-    return {.parse_node = parse_node};
-  }
-
-  auto parse_node_or_invalid() const -> Parse::Node { return parse_node; }
-};
-
-// Base class for nodes that have no `parse_node` field.
-struct HasNoParseNodeBase {
-  static auto FromParseNode(Parse::Node /*parse_node*/) -> HasNoParseNodeBase {
-    return {};
-  }
-
-  auto parse_node_or_invalid() const -> Parse::Node {
-    return Parse::Node::Invalid;
-  }
-};
-
-// ParseNodeBase<T> holds the `parse_node` field if the node has a parse tree
-// node, and is either HasParseNodeBase or HasNoParseNodeBase.
-template <typename T>
-using ParseNodeBase =
-    std::conditional_t<GetWithDefault<T, GetHasParseNode, std::true_type>{},
-                       HasParseNodeBase, HasNoParseNodeBase>;
-
-// Base class for nodes that have a `type_id` field.
-struct HasTypeBase {
-  TypeId type_id;
-
-  static auto FromTypeId(TypeId type_id) -> HasTypeBase {
-    return {.type_id = type_id};
-  }
-
-  auto type_id_or_invalid() const -> TypeId { return type_id; }
-};
-
-// Base class for nodes that have no `type_id` field.
-struct HasNoTypeBase {
-  static auto FromTypeId(TypeId /*type_id*/) -> HasNoTypeBase { return {}; }
-
-  auto type_id_or_invalid() const -> TypeId { return TypeId::Invalid; }
-};
-
-// TypeBase<T> holds the `type_id` field if the node has a type, and is either
-// HasTypeBase or HasNoTypeBase.
-template <typename T>
-using TypeBase =
-    std::conditional_t<GetWithDefault<T, GetHasType, std::true_type>{},
-                       HasTypeBase, HasNoTypeBase>;
-
-// Convert a field from its raw representation.
-template <typename T>
-constexpr auto FromRaw(int32_t raw) -> T {
-  return T(raw);
-}
-template <>
-constexpr auto FromRaw<BuiltinKind>(int32_t raw) -> BuiltinKind {
-  return BuiltinKind::FromInt(raw);
+// Typed nodes can be printed by converting them to nodes.
```

Comment:
Is the indentation weird here?

Reply:
It looks OK to me, but github seems to not deal very well with the large deletions here in the side-by-side view.